### PR TITLE
8266749: AArch64: Backtracing broken on PAC enabled systems

### PR DESCRIPTION
**Pull request by a74nh**
On PAC systems, native code may sign return addresses before saving
them to the stack. We must ensure we strip the any signed bits in
order to walk the stack.
Add extra asserts in places where we do not expect saved return
addresses to be signed.

On non-PAC systems, all PAC instructions are treated as NOPs.

On Apple, use the provided ptrauth interface instead of asm
as the compiler may optimise further.

Fedora 33 compiles all distro packages using PAC. Running the distro
provided OpenJDK-latest in GDB on a PAC system:

Thread 2 "java" hit Breakpoint 1, 0x0000fffff68d7fe4 in init_globals() ()
   from /usr/lib/jvm/java-16-openjdk-16.0.1.0.9-1.rolling.fc33.aarch64-fastdebug/lib/server/libjvm.so
(gdb) call (int)pns($sp, $fp, $pc)

"Executing pns"
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0xe26fe4]  init_globals()+0x10
C  0x006ffffff74750c4
C  0x0042fffff6a7f84c
C  0x0037fffff7fa0954
C  0x0030fffff7fa4540
C  0x0078fffff7d980c8

OpenJDK with this patch at the same breakpoint:

(gdb) call (int)pns($sp, $fp, $pc)
"Executing pns"
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x189c47c]  Threads::create_vm(JavaVMInitArgs*, bool*)+0x27c
V  [libjvm.so+0xf527a0]  JNI_CreateJavaVM+0xc0
C  [libjli.so+0x3860]  JavaMain+0x7c
C  [libjli.so+0x732c]  ThreadJavaMain+0xc
C  [libpthread.so.0+0x80c8]  start_thread+0xd8

OpenJDK with this patch breakpointed at pd_hotspot_signal_handler:

"Executing pns"
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x148a730]  PosixSignals::pd_hotspot_signal_handler(int, siginfo_t*, ucontext_t*, JavaThread*)+0x0
C  [linux-vdso.so.1+0x80c]  __kernel_rt_sigreturn+0x0
J 53 c1 jdk.internal.org.objectweb.asm.SymbolTable.addConstantUtf8(Ljava/lang/String;)I java.base (98 bytes) @ 0x0000ffffe159cc3c [0x0000ffffe159cb40+0x00000000000000fc]
j  jdk.internal.org.objectweb.asm.SymbolTable.setMajorVersionAndClassName(ILjava/lang/String;)I+12 java.base
j  jdk.internal.org.objectweb.asm.ClassWriter.visit(IILjava/lang/String;Ljava/lang/String;Ljava/lang/String;[Ljava/lang/String;)V+20 java.base
j  java.lang.invoke.InvokerBytecodeGenerator.classFilePrologue()Ljdk/internal/org/objectweb/asm/ClassWriter;+30 java.base
j  java.lang.invoke.InvokerBytecodeGenerator.generateCustomizedCodeBytes()[B+1 java.base
j  java.lang.invoke.InvokerBytecodeGenerator.generateCustomizedCode(Ljava/lang/invoke/LambdaForm;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/MemberName;+27 java.base
j  java.lang.invoke.LambdaForm.compileToBytecode()V+69 java.base
j  java.lang.invoke.DirectMethodHandle.makePreparedLambdaForm(Ljava/lang/invoke/MethodType;I)Ljava/lang/invoke/LambdaForm;+792 java.base
j  java.lang.invoke.DirectMethodHandle.preparedLambdaForm(Ljava/lang/invoke/MethodType;I)Ljava/lang/invoke/LambdaForm;+17 java.base
j  java.lang.invoke.DirectMethodHandle.preparedLambdaForm(Ljava/lang/invoke/MemberName;Z)Ljava/lang/invoke/LambdaForm;+163 java.base
j  java.lang.invoke.DirectMethodHandle.preparedLambdaForm(Ljava/lang/invoke/MemberName;)Ljava/lang/invoke/LambdaForm;+2 java.base
j  java.lang.invoke.DirectMethodHandle.make(BLjava/lang/Class;Ljava/lang/invoke/MemberName;Ljava/lang/Class;)Ljava/lang/invoke/DirectMethodHandle;+159 java.base
j  java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(BLjava/lang/Class;Ljava/lang/invoke/MemberName;ZZLjava/lang/invoke/MethodHandles$Lookup;)Ljava/lang/invoke/MethodHandle;+210 java.base
j  java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(BLjava/lang/Class;Ljava/lang/invoke/MemberName;Ljava/lang/invoke/MethodHandles$Lookup;)Ljava/lang/invoke/MethodHandle;+14 java.base
j  java.lang.invoke.MethodHandles$Lookup.getDirectMethodForConstant(BLjava/lang/Class;Ljava/lang/invoke/MemberName;)Ljava/lang/invoke/MethodHandle;+31 java.base
j  java.lang.invoke.MethodHandles$Lookup.linkMethodHandleConstant(BLjava/lang/Class;Ljava/lang/String;Ljava/lang/Object;)Ljava/lang/invoke/MethodHandle;+153 java.base
j  java.lang.invoke.MethodHandleNatives.linkMethodHandleConstant(Ljava/lang/Class;ILjava/lang/Class;Ljava/lang/String;Ljava/lang/Object;)Ljava/lang/invoke/MethodHandle;+38 java.base
v  ~StubRoutines::call_stub
V  [libjvm.so+0xe20118]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, Thread*)+0x5c8
V  [libjvm.so+0xe20f64]  JavaCalls::call_static(JavaValue*, Klass*, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x284
V  [libjvm.so+0x184b778]  SystemDictionary::link_method_handle_constant(Klass*, int, Klass*, Symbol*, Symbol*, Thread*)+0x398
V  [libjvm.so+0xa1f104]  ConstantPool::resolve_constant_at_impl(constantPoolHandle const&, int, int, bool*, Thread*)+0xca0
V  [libjvm.so+0xa1fb6c]  ConstantPool::copy_bootstrap_arguments_at_impl(constantPoolHandle const&, int, int, int, objArrayHandle, int, bool, Handle, Thread*)+0x3fc
V  [libjvm.so+0x6bef6c]  BootstrapInfo::resolve_args(Thread*)+0xcbc
V  [libjvm.so+0x6c1538]  BootstrapInfo::resolve_bsm(Thread*)+0x1194
V  [libjvm.so+0x184d300]  SystemDictionary::invoke_bootstrap_method(BootstrapInfo&, Thread*)+0x30
V  [libjvm.so+0x120450c]  LinkResolver::resolve_dynamic_call(CallInfo&, BootstrapInfo&, Thread*)+0x2c
V  [libjvm.so+0x1204b1c]  LinkResolver::resolve_invokedynamic(CallInfo&, constantPoolHandle const&, int, Thread*)+0x1bc
V  [libjvm.so+0xe0ecc4]  InterpreterRuntime::resolve_invokedynamic(JavaThread*)+0x190
V  [libjvm.so+0xe123a0]  InterpreterRuntime::resolve_from_cache(JavaThread*, Bytecodes::Code)+0x160
j  jdk.internal.module.ModulePath.explodedPackages(Ljava/nio/file/Path;)Ljava/util/Set;+5 java.base
j  jdk.internal.module.ModulePath.lambda$readExplodedModule$9(Ljava/nio/file/Path;)Ljava/util/Set;+2 java.base
j  jdk.internal.module.ModulePath$$Lambda$2+0x000000010003bbe0.get()Ljava/lang/Object;+8 java.base
j  jdk.internal.module.ModuleInfo.doRead(Ljava/io/DataInput;)Ljdk/internal/module/ModuleInfo$Attributes;+762 java.base
j  jdk.internal.module.ModuleInfo.read(Ljava/io/InputStream;Ljava/util/function/Supplier;)Ljdk/internal/module/ModuleInfo$Attributes;+16 java.base
j  jdk.internal.module.ModulePath.readExplodedModule(Ljava/nio/file/Path;)Ljava/lang/module/ModuleReference;+35 java.base
j  jdk.internal.module.ModulePath.readModule(Ljava/nio/file/Path;Ljava/nio/file/attribute/BasicFileAttributes;)Ljava/lang/module/ModuleReference;+11 java.base
j  jdk.internal.module.ModulePath.scanDirectory(Ljava/nio/file/Path;)Ljava/util/Map;+69 java.base
j  jdk.internal.module.ModulePath.scan(Ljava/nio/file/Path;)Ljava/util/Map;+60 java.base
j  jdk.internal.module.ModulePath.scanNextEntry()V+23 java.base
j  jdk.internal.module.ModulePath.find(Ljava/lang/String;)Ljava/util/Optional;+36 java.base
j  jdk.internal.module.SystemModuleFinders$1.lambda$find$0(Ljava/lang/module/ModuleFinder;Ljava/lang/String;)Ljava/util/Optional;+2 java.base
j  jdk.internal.module.SystemModuleFinders$1$$Lambda$1+0x0000000100033b00.run()Ljava/lang/Object;+8 java.base
j  java.security.AccessController.executePrivileged(Ljava/security/PrivilegedAction;Ljava/security/AccessControlContext;Ljava/lang/Class;)Ljava/lang/Object;+29 java.base
j  java.security.AccessController.doPrivileged(Ljava/security/PrivilegedAction;)Ljava/lang/Object;+5 java.base
j  jdk.internal.module.SystemModuleFinders$1.find(Ljava/lang/String;)Ljava/util/Optional;+12 java.base
j  jdk.internal.module.ModuleBootstrap.boot2()Ljava/lang/ModuleLayer;+304 java.base
j  jdk.internal.module.ModuleBootstrap.boot()Ljava/lang/ModuleLayer;+64 java.base
j  java.lang.System.initPhase2(ZZ)I+0 java.base
v  ~StubRoutines::call_stub
V  [libjvm.so+0xe20118]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, Thread*)+0x5c8
V  [libjvm.so+0xe20f64]  JavaCalls::call_static(JavaValue*, Klass*, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x284
V  [libjvm.so+0x189c7bc]  Threads::create_vm(JavaVMInitArgs*, bool*)+0x5bc
V  [libjvm.so+0xf527a0]  JNI_CreateJavaVM+0xc0
C  [libjli.so+0x3860]  JavaMain+0x7c
C  [libjli.so+0x732c]  ThreadJavaMain+0xc
C  [libpthread.so.0+0x80c8]  start_thread+0xd8

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266749](https://bugs.openjdk.java.net/browse/JDK-8266749): AArch64: Backtracing broken on PAC enabled systems

### Reviewers
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to 406eeed551f7591bc0cd603b3752c29c7c69294c
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4029/head:pull/4029` \
`$ git checkout pull/4029`

Update a local copy of the PR: \
`$ git checkout pull/4029` \
`$ git pull https://git.openjdk.java.net/jdk pull/4029/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4029`

View PR using the GUI difftool: \
`$ git pr show -t 4029`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4029.diff">https://git.openjdk.java.net/jdk/pull/4029.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/pauth_aarch64.hpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/pauth_aarch64.hpp:3]: This file is not a derived work of Oracle's, as far as I can see.

**Comment by a74nh** [comment]
Windows build is failing because it doesn't like volatile:
d:\a\jdk\jdk\jdk\src\hotspot\cpu\aarch64\pauth_aarch64.hpp(43): error C2059: syntax error: 'volatile'

I guess the fix is to put pauth_aarch64.hpp into the os_arch/ directories. I wanted to avoid that because each copy would be almost identical. Unless anyone has any better ideas?

**Comment by VladimirKempik** [comment]
let me check with macos/aarch64 on macos11.4b as it was giving problems with return address having pac signature sometimes too.

**Comment by VladimirKempik** [comment]
> let me check with macos/aarch64 on macos11.4b as it was giving problems with return address having pac signature sometimes too.
> 
> Nop, still crashes in SharedRuntime::raw_exception_handler_for_return_address with return address having pac signature in it (probably it's pac, not 100% sure), will investigate further

on macos11.4b it's getting something that looks like PAC-ed pointer in src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp line 487
  address generate_forward_exception() {
...
    __ mov(c_rarg1, lr);  <-- here
not sure how it got here bypassing all of your checks

maybe it's too early as 11.4 is still beta. ( beta3 ) and java_m1 is not arm64e process (arm64e = arm64 + PAC)

**Comment by a74nh** [comment]
> let me check with macos/aarch64 on macos11.4b as it was giving problems with return address having pac signature sometimes too.
> 
> Nop, still crashes in SharedRuntime::raw_exception_handler_for_return_address with return address having pac signature in it (probably it's pac, not 100% sure), will investigate further

I suspect that 

> > let me check with macos/aarch64 on macos11.4b as it was giving problems with return address having pac signature sometimes too.
> > Nop, still crashes in SharedRuntime::raw_exception_handler_for_return_address with return address having pac signature in it (probably it's pac, not 100% sure), will investigate further
> 
> on macos11.4b it's getting something that looks like PAC-ed pointer in src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp line 487
> address generate_forward_exception() {
> ...
> __ mov(c_rarg1, lr); <-- here
> not sure how it got here bypassing all of your checks
> 
> maybe it's too early as 11.4 is still beta. ( beta3 ) and java_m1 is not arm64e process (arm64e = arm64 + PAC)

Apple's Arm64e also signs function pointers (unlike Linux, as it'd break the published AArch64 ABI). If it's only failing in the beta then maybe they are signing additional things. (I've not got access to that).

It might be worth copying the assert I added to the linux_aarch64 version of fetch_compiled_frame_from_context() into the os_bsd version. That might catch what your seeing earlier.   (Happy to add that to my patch too).

File: src/hotspot/cpu/aarch64/pauth_aarch64.hpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/pauth_aarch64.hpp:44]: There's no need for this to be volatile.

**Comment by VladimirKempik** [comment]
> It might be worth copying the assert I added to the linux_aarch64 version of fetch_compiled_frame_from_context() into the os_bsd version. That might catch what your seeing earlier. (Happy to add that to my patch too).

I have tried that and few more things, nothing catches that, only clearing upper 28 bits in c_rarg1 in generate_forward_exception helps. pauth_ptr_is_raw doesn't think it's pac-ed ptr ( maybe it always returns true on arm64 ( not arm64e))

here is example of ptrs on one run

return_address must be a return address: 0x9221800138009a5c
return_address must be a return address: 0x5c08800138009a5c
return_address must be a return address: 0x5c08800138009a5c
return_address must be a return address: 0x8713800138009a5c
return_address must be a return address: 0x0514800138009a5c
return_address must be a return address: 0x8d7a000138009a5c
return_address must be a return address: 0x403c000138009a5c
return_address must be a return address: 0x492d000138009a5c
return_address must be a return address: 0x403c000138009a5c

**Comment by theRealAph** [comment]
> Windows build is failing because it doesn't like volatile:
> d:\a\jdk\jdk\jdk\src\hotspot\cpu\aarch64\pauth_aarch64.hpp(43): error C2059: syntax error: 'volatile'
> 
> I guess the fix is to put pauth_aarch64.hpp into the os_arch/ directories. I wanted to avoid that because each copy would be almost identical. Unless anyone has any better ideas?

Sure, write a version in assembler and call it as extern "C".

File: src/hotspot/cpu/aarch64/pauth_aarch64.hpp
Comment (a74nh) [reply src/hotspot/cpu/aarch64/pauth_aarch64.hpp:3]: As I understood it, as stated in the Oracle Contributor Agreement, we share copyright with Oracle.  And as far as I can tell, Arm has always added new files with both an Arm and Oracle line (most obvious example I can find of this is cpu/aarch64/registerMap_aarch64.cpp, which has a single log entry from Nick).

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/pauth_aarch64.hpp:3]: OK. Your call, but you don't need to do that.

**Comment by lewurm** [comment]
> maybe it's too early as 11.4 is still beta. ( beta3 ) and java_m1 is not arm64e process (arm64e = arm64 + PAC)

I think you have to explicitly turn on arm64e on macOS, see at the bottom here: https://developer.apple.com/documentation/driverkit/debugging_and_testing_system_extensions (I have never tried it myself).

As far as I know Windows/AArch64 doesn't support PAC, so imho we don't need to do anything on Windows there in that regard. I'll ask around internally if there are any plans for PAC on Windows.

**Comment by a74nh** [comment]
* Split pauth_aarch64.hpp into os_arch variants to fix windows build
* * Windows functions are blank because no PAC on windows yet.
* Added assert in bsd signal handler

**Comment by a74nh** [comment]
Any outstanding issues with this?
* The patch fixes issues on Linux distros with PAC support.
* Changes will help Apple arm64e, but there are additional issues that need fixing in a future PR.
* No changes to windows port (all new functions do nothing). If PAC is added to windows, the functions will need filling out.

File: src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp
Comment (gerard-ziemski) [inline src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp:42]: Wouldn't it be safer to use the alias `ptrauth_key_process_dependent_code` instead of the underlaying primitive enum `ptrauth_key_asib`?

  ptrauth_strip(ptr, ptrauth_key_process_dependent_code);

Can you please elaborate on why this particular key was used, for example why not `ptrauth_key_process_independent_code`? There are so many other...

Comment (a74nh) [reply src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp:42]: For the values being stripped in this patch, they will always be return addresses saved on the stack. My first choice here was ptrauth_key_return_address, but I wanted it to be a little more generic and:
ptrauth_key_return_address = ptrauth_key_process_dependent_code = ptrauth_key_asib [1]

For a strip function, the key shouldn't matter as xpaci/xaplri instructions work for both IA and IB keys. However when I've dumped my test binaries on mac, it looked like the compiler used generic bitmasking instead of the xpac instructions. So it's possible the relevant key does matter here. Using the IB key ensures ptrauth_strip() always does the stripping (according to ptrauth.h comments), which is probably the safest option for now. I suspect full mac arm64e support would eventually require having multiple strip functions (pauth_strip_retaddr(), pauth_strip_functptr(), etc)

[1] https://opensource.apple.com/source/xnu/xnu-4903.241.1/EXTERNAL_HEADERS/ptrauth.h.auto.html

Comment (gerard-ziemski) [reply src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp:42]: I don't 100% understand this piece of code on macOS (I stripped non macOS code from it for clarity):

```
inline address pauth_strip_pointer(address ptr) {
   ptrauth_strip(ptr, ptrauth_key_asib);
   return ptr;
 }
```

shouldn't it be something more like:

```
inline address pauth_strip_pointer(address ptr) {
   return ptrauth_strip(ptr, ptrauth_key_asib);
 }
```

The original code compiles with **"Expression result unused"** for me.

Comment (a74nh) [reply src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp:42]: Good spot. I'll fix this.

**Comment by VladimirKempik** [comment]
> > It might be worth copying the assert I added to the linux_aarch64 version of fetch_compiled_frame_from_context() into the os_bsd version. That might catch what your seeing earlier. (Happy to add that to my patch too).
> 
> I have tried that and few more things, nothing catches that, only clearing upper 28 bits in c_rarg1 in generate_forward_exception helps. pauth_ptr_is_raw doesn't think it's pac-ed ptr ( maybe it always returns true on arm64 ( not arm64e))
> 
> here is example of ptrs on one run
> 
> return_address must be a return address: 0x9221800138009a5c
> return_address must be a return address: 0x5c08800138009a5c
> return_address must be a return address: 0x5c08800138009a5c
> return_address must be a return address: 0x8713800138009a5c
> return_address must be a return address: 0x0514800138009a5c
> return_address must be a return address: 0x8d7a000138009a5c
> return_address must be a return address: 0x403c000138009a5c
> return_address must be a return address: 0x492d000138009a5c
> return_address must be a return address: 0x403c000138009a5c
> 
> This is not any suggestion for this PR, just FYI

Turned out it was yet another PAC issue, fixed now for aarch64 (see JDK-8267235)

**Comment by gerard-ziemski** [comment]
Please let me know if I understand things correctly:

Apple's **M1** is **arm64e** architecture, so therefore it is **PAC** architecture, however, currently that's only active in **iOS**.

**macOS** does not currently support **PAC**, but we're adding the support to be ready whenever Apple decides to turn it on?

On **PAC** systems the following code:

`ptrauth_strip(&strcpy, ptrauth_key_asib) == &strcpy`

would be false, and on **non PAC** OS , like for now **macOS** it will always be true (since ptrauth_strip is a NOP macro), is that correct?

**Comment by gerard-ziemski** [comment]
> > Please let me know if I understand things correctly:
> > Apple's **M1** is **arm64e** architecture, so therefore it is **PAC** architecture, however, currently that's only active in **iOS**.
> > **macOS** does not currently support **PAC**, but we're adding the support to be ready whenever Apple decides to turn it on?
> 
> MacOS does support PAC, but it's an experimental feature that needs enabling at machine boot up.
> For example, all the standard tools in /bin are built for arm64e, and when they are run, the PAC instructions are treated as NOPs.
> 
> Apple could turn on PAC as default at any time, and any app compiled for arm64e will suddenly be PAC protected. I suspect we'll have a little warning first. Given that I was adding the fix for Linux, it made sense to also proactively fix MacOS at the same time.
> 
> > On **PAC** systems the following code:
> > `ptrauth_strip(&strcpy, ptrauth_key_asib) == &strcpy`
> > would be false, and on **non PAC** OS , like for now **macOS** it will always be true (since ptrauth_strip is a NOP macro), is that correct?
> 
> Assuming that &strcpy is a signed pointer, then yes - it will be false on a PAC enabled OS.
> If the pointer isn't signed or PAC is disabled, then it will be true.

So currently, there is no way, by default, to test PAC on macOS. Is it turned on by default on Linux arrch64 starting at some OS version, or is it optional like macOS?

Did you give any thoughts to writing gtests that exercise these new PAC related functionality?

**Comment by gerard-ziemski** [comment]
> > MacOS does support PAC, but it's an experimental feature that needs enabling at machine boot up.

I can only seem to find info on how to enable/disable System Integrity Protection, how exactly do you do it for PAC?

**Comment by VladimirKempik** [comment]
Hello
I don't think that's correct description of current PAC state on macos11
if binary is of type arm64e then it has pac enabled and it works. (most of system binaries are already arm64e, also the kernel and kernel extensions are arm64e already)
To be able to test your own arm64e binary one need to add to boot-args ( https://developer.apple.com/documentation/driverkit/debugging_and_testing_system_extensions?language=objc ) -arm64e_preview_abi

But, there is one thing, the shared library cache is arm64e only ( check folder /System/Library/dyld), and all of  few remaining dylibs ( check any in /usr/lib) is arm64e only ( plus x86_64).
All of these arm64e libs use PAC and it works ( not disabled/NOP-ed), as a result we had a bug recently - https://bugs.openjdk.java.net/browse/JDK-8267235

jvm probably needs to sanitize any pointer coming from native libs.

**Comment by gerard-ziemski** [comment]
I see this pattern repeating:

```
  address sender_pc = this->sender_pc();
   // Interpreter should never sign the return address.
   assert(pauth_ptr_is_raw(sender_pc), "cannot be signed");

   return frame(sender_sp, unextended_sp, link(), sender_pc);
```

Why can't we just add the **assert** to the **frame()** constructor instead?

File: src/hotspot/cpu/aarch64/frame_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/frame_aarch64.cpp:516]: Would it not make more sense to strip the `sender_pc()` ? Don't all users expect it to be stripped?

**Comment by a74nh** [comment]
> Why can't we just add the **assert** to the **frame()** constructor instead?

> Would it not make more sense to strip the `sender_pc()` ? Don't all users expect it to be stripped?

I'll try both these see which is the better option (one probably counteracts the other).


> I don't think that's correct description of current PAC state on macos11
> if binary is of type arm64e then it has pac enabled and it works. (most of system binaries are already arm64e, also the kernel and kernel extensions are arm64e already)
> To be able to test your own arm64e binary one need to add to boot-args ( https://developer.apple.com/documentation/driverkit/debugging_and_testing_system_extensions?language=objc ) -arm64e_preview_abi, otherwise the kernel won't allow custom arm64e app to start.
> 
> But, there is one thing, the shared library cache is arm64e only ( check folder /System/Library/dyld), and all of few remaining dylibs ( check any in /usr/lib) is arm64e only ( plus x86_64). arm64 apps use arm64e libs.
> All of these arm64e libs use PAC and it works ( not disabled/NOP-ed), as a result we had a bug recently - https://bugs.openjdk.java.net/browse/JDK-8267235
> 
> jvm probably needs to sanitize any pointer coming from native system libs.

Much better description than I gave, thanks :)

File: src/hotspot/cpu/aarch64/frame_aarch64.cpp
Comment (a74nh) [reply src/hotspot/cpu/aarch64/frame_aarch64.cpp:509]: > Would it not make more sense to strip the `sender_pc()` ? Don't all users expect it to be stripped?

The current code is being fairly cautious:

Native frames: saved PC might be signed. Therefore strip the pc.

Interpreter frames and compiled frames: the PC should never be signed. Assert this is the case.

For both Interpreter and Native frames we are getting the pc via sender_pc(), so can't push the strip inside sender_pc().

Alternatively, we could always strip instead of asserting. But that's not quite as nice.

**Comment by a74nh** [comment]
> I see this pattern repeating:
> 
> ```
>    address sender_pc = this->sender_pc();
>    // Interpreter should never sign the return address.
>    assert(pauth_ptr_is_raw(sender_pc), "cannot be signed");
> 
>    return frame(sender_sp, unextended_sp, link(), sender_pc);
> ```
> 
> Why can't we just add the **assert** to the **frame()** constructor instead?

I've made this change.

File: src/hotspot/cpu/aarch64/frame_aarch64.cpp
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/frame_aarch64.cpp:509]: Mmm, but you very easily fix that by defining `raw_sender_pc()`. Most users don't want to be bothered by this stuff, and having `sender_pc()` return an encrypted PC is, frankly, maintainer abuse. Anyone who wants the `raw_sender_pc()` should ask for it.

Comment (a74nh) [reply src/hotspot/cpu/aarch64/frame_aarch64.cpp]: Ok, that way makes sense. Updated as suggested.

File: src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp
Comment (gerard-ziemski) [inline src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp:176]: Do we need this, now that we have the assert in frame constructor ?

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp
Comment (gerard-ziemski) [inline src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:145]: Do we need this, now that we have the assert in frame constructor ?

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp
Comment (a74nh) [reply src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp]: Removed as suggested.

Also fixed some build failures.

File: src/hotspot/cpu/aarch64/pauth_aarch64.hpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/pauth_aarch64.hpp:31]: This predicate seems to me to be be misnamed: it's checking for unsigned/stripped, not for raw. The raw value is whatever gets saved in LR/pushed onto the stack.

Comment (a74nh) [reply src/hotspot/cpu/aarch64/pauth_aarch64.hpp:31]: Ah, yes, because the name pauth_ptr_is_raw() now clashes with sender_pc_raw().

I'll fix up, one way or the other. Don't really want to call it pauth_ptr_is_unsigned() or pauth_ptr_is_authenticated(), because they seem to imply different things at first glance.

Comment (a74nh) [reply src/hotspot/cpu/aarch64/pauth_aarch64.hpp:31]: raw is generally accepted as the correct name for a pointer that has not been signed (or has been authenticated/stripped). 
For example see: https://github.com/apple/llvm-project/blob/a63a81bd9911f87a0b5dcd5bdd7ccdda7124af87/clang/docs/PointerAuthentication.rst#basic-concepts

So, pauth_ptr_is_raw() is the correct name.

But, now what to rename sender_pc_raw() to ?  :)
The PAC coder would say:
   inline address  frame::sender_pc()       const { return *sender_pc_addr(); }
   inline address  frame::sender_pc_authenticated()           const { return pauth_strip_pointer(sender_pc()); }
But that's not right here for a common interface.

How about:
   inline address  frame::stored_sender_pc()       const { return *sender_pc_addr(); }
   inline address  frame::sender_pc()           const { return pauth_strip_pointer(stored_sender_pc()); }

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/pauth_aarch64.hpp:31]: Off the top of my head, I would have thought that it makes the most sense to use Arm's terminology, which is "signed pointer". So, that'd be `frame::signed_sender_pc()` or `frame::maybe_signed_sender_pc()`.

**Review by gerard-ziemski — APPROVED** [review]

File: src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp
Comment (theRealAph) [inline src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp:46]: One minor nit: I wonder if it's necessary to do those two MOVs. We could put ptr into x30 thusly:

```
  register uint64_t result __asm__("x30");
  result = ptr;
  asm (XPACLRI : "+r"(result));
  return result;
```
This would generate fewer MOVs in most cases, but as MOV instructions don't even issue in M1 it's perhaps not worth bothering. Up to you.

Comment (a74nh) [reply src/hotspot/os_cpu/bsd_aarch64/pauth_bsd_aarch64.inline.hpp]: >One minor nit: I wonder if it's necessary to do those two MOVs. We could put ptr into x30 thusly:

Looks much better like this. Updated.

**Review by theRealAph — APPROVED** [review]
I don't think anyone can improve on that. Thanks for your patience.

**Pull request closed without merging**
